### ISSUE TITLE
New Rule Proposal: `no-unnecessary-props`

### ISSUE THREAD

**Issue by karlhorky** [yannickcr/eslint-plugin-react#2996]
Hi there @ljharb and others  👋

One pattern that I see a lot in beginner student code is passing props unnecessarily (before they have learned TypeScript or used prop types).

Maybe this would be a nice rule to have generally: `no-unnecessary-props` - detect prop values which have been passed but not used in component.

I'm not sure if tracking whether the values are being used is simple in ESLint.

This is not to be confused with [`no-unused-prop-types`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unused-prop-types.md), which as far as I can tell only has to do with prop types.

Example of **incorrect** code:

```jsx
function Input (props) {
  return (
    <input
      type={props.type}
      // props.handleChange unused in component function
    />
  );
}

<Input
  type="text"
  handleChange={() => {}} // ESLint problem here
/>
```

Examples of **correct** code:

```jsx
function Input (props) {
  return (
    <input
      type={props.type}
      onChange={props.handleChange}
    />
  );
}

<Input
  type="text"
  handleChange={() => {}}
/>
```

**Comment by ljharb** [comment]
This simply isn’t achievable without type information.

I suggest people use https://npmjs.com/prop-types-exact for this, and then run their unit tests.

**Comment by karlhorky** [comment]
Ah, too bad :(  Are there any other rules that use type information if available (eg. if [`typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint) is set up)?

Would this be possible / desirable for `eslint-plugin-react`? I could imagine this would enable a bunch of different use cases...

I guess the `typescript-eslint` packages are not focused on React rules, so I guess they wouldn't accept such an issue over there... 🤔

**Comment by ljharb** [comment]
TypeScript already does this for typed components automatically.

**Comment by karlhorky** [comment]
Yeah I know, I'm talking about untyped (inferred) components with JS though (`allowJs` flag)

**Comment by ljharb** [comment]
With the allowJs flag, typescript does that automatically for js files too. Have you tried it?

**Comment by karlhorky** [comment]
Yep, all student projects use it: https://github.com/upleveled/eslint-config-upleveled/blob/main/templates/tsconfig.json#L15

Since we're teaching beginners, we don't use the `checkJs` flag, which would yield the behavior that you're describing (but also lead to a lot of other type errors)

**Comment by ljharb** [comment]
ah, fair enough. i think that flag is exactly what you’re asking for here - without type information, we couldn’t do much better.

**Comment by karlhorky** [comment]
> without type information, we couldn’t do any better

Yeah, that's what I'm saying though: with `allowJs` (even without `checkJs`), there is type information available - that's how we [enable the fancy rules from `typescript-eslint`](https://github.com/upleveled/eslint-config-upleveled/blob/037a847c0199bd2eb8f536e9de87d520f0c33bfa/index.cjs#L143-L149), even at the early stage with these beginners.

**Comment by karlhorky** [comment]
But yeah, if `eslint-plugin-react` is not amenable to adding functionality based on type information (or the use case is too narrow), then we can just leave it at that.

**Comment by ljharb** [comment]
It's not about "amenable" - it's just not an option. Since the vast majority of components are used in a file different from the one in which they're defined, we'd have to parse the entire project to get that information - except for a few eslint-plugin-import rules, that is simply never done in the eslint ecosystem.

**Comment by karlhorky** [comment]
I understand that, I'm not proposing that this or any other ESLint plugin read across multiple files, that would be a bunch of work.

I'm proposing that in the case that there is TS type information, that `eslint-plugin-react` use that type information, which would enable not only this feature for `allowJs` users, but also other features. Currently, `allowJs: true, checkJs: false` users have type information, but it is not used to provide TypeScript errors. 

I'm guessing `eslint-plugin-react` would then need to gain access to this information somehow - eg. in a way similar to how `typescript-eslint` does it (maybe even with one of the utilities over there).

**Comment by ljharb** [comment]
There isn't TS type information unless typescript is installed and itself runs on the entire project (which is what typescript-eslint does) - and the very reason you don't enable `checkJs` is the reason it wouldn't likely be useful, since the types wouldn't be accurate enough.

**Comment by karlhorky** [comment]
> it wouldn't likely be useful, since the types wouldn't be accurate enough

Hm, interesting point... 🤔 Although I'm not certain this is true - in practice, it seems like we're getting enough information through inference and DefinitelyTyped libs like `@types/react` to enable multiple `typescript-eslint` rules that require type information. So seems to be working!
